### PR TITLE
feat: implement `requiredStatusChecks` for GitHub and GitLab

### LINKED ISSUE #1853: Implement requiredStatusChecks
#### This is a:

* [ ] Bug report (non-security related)
* [x] Feature request
* [ ] I'm not sure which of those it is

The `requiredStatusChecks` option is not fully implemented - it only works if you set it to `null` (i.e. NO required status checks).

We should implement the feature fully so that users can specify a list of status checks that are required for automerge to be "green" and we ignore the rest.

### PR DESCRIPTION
**Pull request by sarunint**
Closes #1853 (partially)

### CODE REVIEW

**Comment by rarkins** [comment]
There are some obvious use cases here but also some less obvious.

1. If all required status checks are present and all passing: succeeded
2. If any required status checks are present and failed: failed
3. If only some of the status checks are present but all that are present are passing: pending?

The key question is: what to do if the user configures a required status check that fails to show up? I think we should probably be pending forever. What do you think?

**Comment by sarunint** [comment]
Ha, I totally missed those cases.

**Comment by sarunint** [comment]
Let me think for a while.

Also, should I do this for other platforms in this PR? Or should I separate them up?

**Comment by rarkins** [comment]
Ideally we would implement this for all platforms that could support it. It's OK to note a feature is "platform X only" but it can confuse users even still, so ideally we don't do that.

**Comment by sarunint** [comment]
@rarkins I ended up implement for both GitHub and GitLab.

**Comment by rarkins** [comment]
@leipert could you glance at the GitLab part of this?

**Review by viceice — COMMENTED** [review]
Can we extract some util function for this? I see a lot duplicated code.

File: lib/platform/github/index.ts [hunk 1]
```diff
@@ -629,17 +629,44 @@ export async function getBranchStatus(
...
+
   if (checkRuns.length === 0) {
-    return commitStatus.state;
+    return filteredStatusesCombined;
   }
+
   if (
-    commitStatus.state === 'failed' ||
+    filteredStatusesCombined === 'failed' ||
     checkRuns.some(run => run.conclusion === 'failed')
```
[the remark this answers is not in the document]
Comment (sarunint) [reply lib/platform/github/index.ts]: I'll see what I can do.

**Review by rarkins — CHANGES_REQUESTED** [review]
Check runs (github) support. Also, possible to de-duplicate the core logic between github and gitlab implementations?

**Comment by ViceIce** [comment]
@sarunint any news? Are you planning any changes?

**Comment by ViceIce** [comment]
This should be simple adaptable for `bitbucket` and `bitbucket-server`. 

azure would be more difficult.

https://github.com/renovatebot/renovate/blob/4adb0c00527ec3e25a4629ef15e8684663caf94b/lib/platform/azure/index.ts#L353

`branch.commit.statuses`

https://docs.microsoft.com/en-us/rest/api/azure/devops/git/statuses/list?view=azure-devops-rest-5.1
https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20request%20statuses/list?view=azure-devops-rest-5.1

**Comment by sarunint** [comment]
@ViceIce Sorry, I've been very busy this month. I'll try to fix things up this weekend.

**Comment by ViceIce** [comment]
aditional azure info

**GitBranchStatus** interface
https://github.com/microsoft/azure-devops-node-api/blob/d4b692556db58a418e5d7c2b01fb68978746b3b4/api/interfaces/GitInterfaces.ts#L736-L760

**GitCommitRef** interface
https://github.com/microsoft/azure-devops-node-api/blob/d4b692556db58a418e5d7c2b01fb68978746b3b4/api/interfaces/GitInterfaces.ts#L804-L864

**GitStatus** interface
https://github.com/microsoft/azure-devops-node-api/blob/d4b692556db58a418e5d7c2b01fb68978746b3b4/api/interfaces/GitInterfaces.ts#L2456-L2496

**Review by viceice — CHANGES_REQUESTED** [review]
Needs deconflicting

**Pull request closed without merging**